diff --git a/src/packaging/utils.py b/src/packaging/utils.py
index 23450953d..2246e9a06 100644
--- a/src/packaging/utils.py
+++ b/src/packaging/utils.py
@@ -35,7 +35,7 @@ class InvalidSdistFilename(ValueError):
 
 # Core metadata spec for `Name`
 _validate_regex = re.compile(
-    r"^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$", re.IGNORECASE
+    r"^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])\Z", re.IGNORECASE
 )
 _canonicalize_regex = re.compile(r"[-_.]+")
 _normalized_regex = re.compile(r"^([a-z0-9]|[a-z0-9]([a-z0-9-](?!--))*[a-z0-9])$")
diff --git a/tests/test_utils.py b/tests/test_utils.py
index a8cc8d71e..d19776a5b 100644
--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -37,10 +37,22 @@ def test_canonicalize_name(name, expected):
     assert canonicalize_name(name) == expected
 
 
-def test_canonicalize_name_invalid():
+@pytest.mark.parametrize(
+    ("name", "expected"),
+    [
+        ("_not_legal", "-not-legal"),
+        ("hi\n", "hi\n"),
+        ("\nhi", "\nhi"),
+        ("h\ni", "h\ni"),
+        ("hi\r", "hi\r"),
+        ("\rhi", "\rhi"),
+        ("h\ri", "h\ri"),
+    ],
+)
+def test_canonicalize_name_invalid(name, expected):
     with pytest.raises(InvalidName):
-        canonicalize_name("_not_legal", validate=True)
-    assert canonicalize_name("_not_legal") == "-not-legal"
+        canonicalize_name(name, validate=True)
+    assert canonicalize_name(name) == expected
 
 
 @pytest.mark.parametrize(
